### ISSUE TITLE
Changing the start/end time of a Challenge

### ISSUE THREAD

**Issue by juliojj** [codalab/codabench#1155]
Hi, 
I am trying to create a challenge with a dev and a test phase, and they are automatically created to start/end at 02:00 CEST. This generates a conflict in case I set dev and test phases to start at the same day but at different intervals (i.e., Phase "Test" must start after phase "Development" ends). If I set them two start at two different days, both will start/end at 02:00 CEST, and I may lose some hours. Is there any way to edit the start/end time of a challenge on Codabench?
Best 

# TODO

- [x] ~Allow phase 1 end and phase 2 start dates to be same~ (actually it's possible already)
- [x] Fix the interface so we don't need to click again on the dates to hide the error message
Solved by: #1163 
- [x] **Phase start/end time**: Adding the possibility to choose an exact time, instead of just a date. That can be useful to have strict deadline, such as midnight anywhere on earth, etc.
Solved by: #1616

**Comment by Didayolo** [comment]
> This generates a conflict in case I set dev and test phases to start at the same day but at different intervals

Do you mean that both phases *start* at the same date, or dev phase *end* and test phase *start* at this date?

**Comment by juliojj** [comment]
For example, if I set dev phase to end at day 16, and test phase to start at day 16, I will get an error: <Phase "Test" must start after phase "Development" ends>, as I cannot set the time they end/start. 

On the other hand, if I set dev phase to end at day 16, and the test phase to start at day 17, the dev phase will close at 02:00 CET of day 16, and the test phase will start at 02:00 of day 17 (and if I am not wrong, participants will lose 1 24h). That is, it seems all phases start/end at a fixed time (02:00 CET).

Sorry if I made any mistake.

**Comment by juliojj** [comment]
First, I uploaded a bundle of the challenge, for each track (https://www.codabench.org/competitions/1287, and this one https://www.codabench.org/competitions/1290/). Then, I tried to edit the start/end days of dev/test phases. I got the error ("X" must start after phase "Y") when setting the test phase to start the same day as the end of the dev phase. For some reason, I am not getting the error after your last message. Have you fixed anything? In any case, I was wondering whether it is possible to edit the start/end time for each phase, as it seems to be set to 02:00 CET as default.
Thanks

**Comment by Didayolo** [comment]
I did not change anything. But I have also seen this bug sometimes I think, even if I am not able to reproduce it. Let's keep an eye on it.

> In any case, I was wondering whether it is possible to edit the start/end time for each phase, as it seems to be set to 02:00 CET as default.

Right now you can't select a precise time, only the date. That is an interesting feature suggestion.

**Comment by juliojj** [comment]
I see your point ihsaan-ullah, but this generates another problem. For instance, if phase 1 ends on Sep 13 and phase 2 starts on Sep 14, both will end (phase 1) and start (phase 2) at 02:00 CET, according to my tests. That is, the start/end time are predefined. This means that from Sep 13 02:00 to Sep 14 02:00 the competition will be locked. Apart from this, if I want to end the challenge on the last hour of day 15, lets say, I will need to set up it to end on 02:00 CET of day 16, as one cannot control the start/end time.

**Comment by ihsaan-ullah** [comment]
> I see your point ihsaan-ullah, but this generates another problem. For instance, if phase 1 ends on Sep 13 and phase 2 starts on Sep 14, both will end (phase 1) and start (phase 2) at 02:00 CET, according to my tests. That is, the start/end time are predefined. This means that from Sep 13 02:00 to Sep 14 02:00 the competition will be locked. Apart from this, if I want to end the challenge on the last hour of day 15, lets say, I will need to set up it to end on 02:00 CET of day 16, as one cannot control the start/end time.

Now this seems like a bug. We will look into it.

**Comment by Didayolo** [comment]
> If you select the same date (e.g. Septeber 13 2023) as the end data of Phase 1 and start date of Phase 2, you will get this error: Phase 2 must start after Phase 1 ends

Maybe this simply should not return an error. Because, at the upload, it does not generate an error. So, if you just edit the date but let the same date ("16 october 2023" --> "16 october 2023"), then the bug appears, even if the competition was already saved with this setting.

**Comment by ihsaan-ullah** [comment]
> > If you select the same date (e.g. Septeber 13 2023) as the end data of Phase 1 and start date of Phase 2, you will get this error: Phase 2 must start after Phase 1 ends
> 
> Maybe this simply should not return an error. Because, at the upload, it does not generate an error. So, if you just edit the date but let the same date ("16 october 2023" --> "16 october 2023"), then the bug appears, even if the competition was already saved with this setting.

I think it should return error even during the upload. I will check it. 

What we expect is the following:
- phase 2 should start after phase 1
- same dates phase 1 end and phase 2 start are not acceptable
- phase should end at 23:59 and the next should start at 00:00

**Comment by Didayolo** [comment]
> I think it should return error even during the upload.

Why? I think when the end dates are not defined in the `competition.yaml`, they are automatically set to the start date of the next phase. I don't see why we would not allow end and start dates to be the same, it is what is the most commonly wanted as competition organizer. This way, the next phase starts exactly when the previous ends.

**Comment by ihsaan-ullah** [comment]
> Why? 
Becasuse right now we don't have time, just dates. End phase 2 start should be after phase 1 end 

> Ithink when the end dates are not defined in the `competition.yaml`, they are automatically set to the start date of the next phase. 

No, it is set to start date -1 
 
> I don't see why we would not allow end and start dates to be the same, it is what is the most commonly wanted as competition organizer. This way, the next phase starts exactly when the previous ends.

This is actually the bug. If let say phase 1 ends on September 14 and phase 2 starts on september 15 then end time should be September 14 23:59 and start date should be September 15 00:00

**Comment by Didayolo** [comment]
But then you'll have a problem of backward compatibility because it was allowed on CodaLab. I don't see the problem in accepting dates that are greater or equal, instead of strictly greater.

**Comment by ihsaan-ullah** [comment]
I have added a todo in the main comment

**Comment by juliojj** [comment]
Do you mind letting me know when you have a solution for this so that I can edit our two challenges and have the end of dev phases and the start of test phases fixed? Thanks

**Comment by Didayolo** [comment]
OK, after some tests, I realized that:

- You can actually set the same date for phase 1 end and phase 2 start.
- The problem is within the interface. When the error messages shows, you can go inside the settings of the phases, click again to set the same date (so you click again but do not change anything) and the problem is solved. I don't know how to solve this exactly. 

By the way, in the code, you can see that the error is supposed to be returned only if `end` is strictly superior to `start`:

```javascript
                _.forEach(_.range(self.phases.length), i => {
                    if (i !== 0) {
                        let end = Date.parse(self.phases[i - 1].end)
                        let start = Date.parse(self.phases[i].start)
                        if (end > start || !end) {
                            let message = `Phase "${_.get(self.phases[i], 'name', i + 1)}" must start after phase "${_.get(self.phases[i - 1], 'name', i)}" ends`
                            if (!self.warnings.includes(message)) {
                                self.warnings.push(message)
                                self.update()
                            }
                            is_valid = false
                        }
                    }
```

**Comment by Didayolo** [comment]
> Do you mind letting me know when you have a solution for this so that I can edit our two challenges and have the end of dev phases and the start of test phases fixed (as well as the end of the challenge, which is also set to end at 02:00 CET)? Thanks

As you can see in my message above, you can actually set the same time for phase 1 ending and phase 2 starting. You simply need to set it manually for both phases before saving.

**Comment by ihsaan-ullah** [comment]
@Didayolo if you select the same date for both phases, are you able to submit to both at the same time?

**Comment by Didayolo** [comment]
@ihsaan-ullah I just tried and it worked as excepted:

- If I put both dates to 18th of September, then the phase 1 is closed and the phase 2 is open
- If I put both dates to 19th of September, then the phase 1 is open and the phase 2 is closed

I think it just switch at midnight (server time)

**Comment by ihsaan-ullah** [comment]
> @ihsaan-ullah I just tried and it worked as excepted:
> 
> * If I put both dates to 18th of September, then the phase 1 is closed and the phase 2 is open
> * If I put both dates to 19th of September, then the phase 1 is open and the phase 2 is closed
> 
> I think it just switch at midnight (server time)

This is great. I think the problem is only with UI. What about the problem of phase ending at 2 PM or something.

**Comment by Didayolo** [comment]
> What about the problem of phase ending at 2 PM or something.

I added another TODO in the main comment about that

**Comment by juliojj** [comment]
I understood I could edit and save the challenge, eventualy get the error, click on edit and try again and have the problem "solved". If this works, ok, but it would be nice if you could solve this bug so that other organizers don't save the same problem. About the timing, it would be amazing if organisers could have more freedom to setup the start/end time of each phase, as one have in Codalab, instead of having it fixed to start/end at 02:00 CET. Thanks

**Comment by ihsaan-ullah** [comment]
@Didayolo 

To verify if everything else is ok

# Bundle upload

### `V2 bundle`  
When Phase 1 end and Phase 2 start dates are same:
<img width="727" alt="Screenshot 2023-09-18 at 10 52 18 PM" src="https://github.com/codalab/codabench/assets/13259262/fd77c811-e694-42c4-a958-d6a1d6420c3c">

When Phase 1 has no end date
<img width="647" alt="Screenshot 2023-09-18 at 10 53 02 PM" src="https://github.com/codalab/codabench/assets/13259262/31c316fd-9158-4e2c-a972-9cc0831f2123">


### `V1 bundle`
We actually don't have phase end dates so when you upload a bundle with 2 phases so 
Phase 1 end date is set to Phase 2 start date -1
<img width="368" alt="Screenshot 2023-09-18 at 10 56 16 PM" src="https://github.com/codalab/codabench/assets/13259262/14910e3d-6b4b-4a6a-a4e2-81a5ee99ddba">
<img width="323" alt="Screenshot 2023-09-18 at 10 56 21 PM" src="https://github.com/codalab/codabench/assets/13259262/732e0132-96c1-4af7-850a-9e5bff10f609">



# Editor

### `Phase 1`
I am able to set Phase 1 end date as Phase 2 start date

<img width="335" alt="Screenshot 2023-09-18 at 10 59 12 PM" src="https://github.com/codalab/codabench/assets/13259262/d7a8d8a0-c6c7-45b3-9ba9-ef7d12e9e5bf">
<img width="326" alt="Screenshot 2023-09-18 at 10 59 25 PM" src="https://github.com/codalab/codabench/assets/13259262/770b03c8-bb54-4592-8114-b7b8beb5ba6b">

### `Phase 2`
When I set Phase 2 start date to same as phase 1 end date. I see this which is the problem
<img width="966" alt="Screenshot 2023-09-18 at 11 01 51 PM" src="https://github.com/codalab/codabench/assets/13259262/85d243ff-fe7f-46f5-8a85-5b65850b7c75">

You said this:

> The problem is within the interface. When the error messages shows, you can go inside the settings of the phases, click again to set the same date (so you click again but do not change anything) and the problem is solved. I don't know how to solve this exactly.

The message will disappear if you click end date of Phase 1 and do not change anything but will not disappear when you click phase 2 start date

**Comment by Didayolo** [comment]
@juliojj 

Ihsan fixed the problem, and the fix is now merged into develop. It will soon be deployed in the main platform (~ in 1 or 2 weeks).

The next TODO, indeed, is being able to set a precise time. We can keep the 2am by default and add an option to specify another time.

**Comment by nicomy** [comment]
Hello, 
It seems that we can already set a precise time using a yaml bundle V2. Nonetheless, I am unable to change the Timezone. 

Both syntax works : 
![image](https://github.com/user-attachments/assets/f5750d5f-235f-49e6-b8fe-cbe5a3982d38)

![image](https://github.com/user-attachments/assets/7bfc590b-acde-4614-b239-856844a3998f)
 
They are, however, in the UTC time zone. 

Maybe you already knew, however the Wiki should be updated and this thread closed ? 
Or am I  missing something :)

**Comment by Didayolo** [comment]
So we have it using YAML but not the editor.
